packages/compiler/src/compiler/compile.ts

```diff
@@ -170,383 +96,98 @@ export class Compile {
-
-          const refConversation = await refPromptCompile.runWithoutGrouping()
-          this.messages.push(...refConversation.messages)
-          this.accumulatedText = refPromptCompile.accumulatedText
-          this.accumulatedContent = refPromptCompile.accumulatedContent
-          this.accumulatedToolCalls = refPromptCompile.accumulatedToolCalls
-        } catch (error: unknown) {
-          if (error instanceof CompileError) throw error
-          this.baseNodeError(errors.referenceError(error), node)
-        }
-        return
-      }
-
-      this.baseNodeError(errors.unknownTag(node.name), node)
-    }
-
-    //@ts-ignore - Linter knows this should be unreachable. That's what this error is for.
-    this.baseNodeError(errors.unsupportedBaseNodeType(node.type), node)
-  }
-
-  private buildMessage({
-    node,
-    role,
-    attributes,
-    content,
-    toolCalls,
-  }: {
-    node?: MessageTag
-    role: MessageRole
-    attributes: Record<string, unknown>
-    content: MessageContent[]
-    toolCalls: ToolCallReference[]
-  }): Message {
-    if (role !== MessageRole.assistant) {
-      toolCalls.forEach(({ node: toolNode }) => {
-        this.baseNodeError(errors.invalidToolCallPlacement, toolNode)
-      })
-    }
-
-    if (role === MessageRole.system) {
-      return {
-        role,
-        content,
-      } as SystemMessage
-    }
-
-    if (role === MessageRole.user) {
-      return {
-        role,
-        name: attributes.name ? String(attributes.name) : undefined,
-        content,
-      } as UserMessage
-    }
-
-    if (role === MessageRole.assistant) {
-      return {
-        role,
-        toolCalls: toolCalls.map(({ value }) => value),
-        content,
-      } as AssistantMessage
-    }
-
-    if (role === MessageRole.tool) {
-      if (attributes.id === undefined) {
-        this.baseNodeError(errors.toolMessageWithoutId, node!)
-      }
-
-      return {
-        role,
-        id: String(attributes.id),
-        content,
-      } as ToolMessage
-    }
-
-    this.baseNodeError(errors.invalidMessageRole(role), node!)
+  }: ResolveBaseNodeProps<TemplateNode>): Promise<void> {
+    const context: CompileNodeContext<TemplateNode> = {
+      node,
+      scope,
+      isInsideMessageTag,
+      isInsideContentTag,
+      resolveBaseNode: this.resolveBaseNode.bind(this),
+      resolveExpression: this.resolveExpression.bind(this),
+      baseNodeError: this.baseNodeError.bind(this),
+      expressionError: this.expressionError.bind(this),
+      addMessage: this.addMessage.bind(this),
+      addStrayText: this.addStrayText.bind(this),
+      popStrayText: this.popStrayText.bind(this),
+      groupStrayText: this.groupStrayText.bind(this),
+      addContent: this.addContent.bind(this),
+      popContent: this.popContent.bind(this),
+      groupContent: this.groupContent.bind(this),
+      addToolCall: this.addToolCall.bind(this),
+      popToolCalls: this.popToolCalls.bind(this),
+    }
```

Comment:
The context with all required functions to read/modify the global state from the new external methods, as explained in the PR description.